### PR TITLE
2FA Removal Feature  🏷 

### LINKED ISSUE #1693: Create table, model, and job for TwoFactorAuthenticationRemovals 🔧
This is a foundational item for the `2FA Removal Feature`

To complete this item, create the file structure and migrations for 
`TwoFactorAuthenticationRemovals` and `TwoFactorAuthenticationRemovalsJob`.

Also add `UI` to the landing page for `Lost your 2fa?` which directs to an e-mail input. 

See #1652 for the responsibilities of these objects

### PR DESCRIPTION
**Pull request by dlipeles**
### 2FA Removal Feature 🔐 
Group

#### Features 
This pull request fulfills the `2fa Removal Spec` for publishers who have lost their second factor authenticator, but wish to regain access to their account. 

* Options - Provides an alternative to creating a new account and transferring channels.
* Secure Flow - Enforces a waiting period & requires re-verification of channels/wallet.
* Self Service - Publishers can initiate the removal at their convenience.

#### Notes
* Admins can see the progress of a 2fa removal from the publisher page.
* There is a `2 week waiting period` before 2fa removal , then a `4 week lockout` on payments for the publisher's security.
* Admins can preempt a 2fa removal or lockout from the admin dashboard.

#### Setup
* Set the `TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD` to `300` seconds and `LOCKED_STATUS_WAITING_PERIOD` to `300` seconds. 
* Set the cron job in sidekiq to every minute `* * * * *`.
* Create a publisher with `2fa enabled`, `channels`, and a `wallet`.
* Request the 2fa removal from the landing page, confirm the e-mail. 

#### In Action
* Log in as an `Admin` to see progress on the job. 
* You should see the following sequence: 
1. `2fa job registered` 
2. `Reminder e-mails sent`
3. `Publisher 2fa removed`, `Publisher channels removed`, `Publisher wallet disconnected`, `Publisher state becomes LOCKED`
4. `2fa job completed`, `Publisher state becomes ACTIVE`

### CODE REVIEW

**Comment by corymcdonald** [comment]
I wonder if it should be more streamlined? Like after the user logs in and they ca't get passed the 2 FA screen we already have their email so I don't know if the user necessarily needs to enter their email?

File: app/controllers/publishers_controller.rb [hunk 1]
```diff
@@ -199,6 +203,17 @@ def update
     end
   end
 
+  def request_two_factor_authentication_removal
+    publisher = Publisher.find_by_email(params[:email])
+    if publisher
+      flash[:notice] = t("publishers.two_factor_authentication_removal.request_success")
+      MailerServices::TwoFactorAuthenticationRemovalRequestEmailer.new(publisher: publisher).perform
+    else
+      flash[:warning] = t("publishers.two_factor_authentication_removal.request_not_found")
```
Comment (yachtcaptain23) [inline app/controllers/publishers_controller.rb]: This exposes the email addresses that we have. We should change the message to the same as the above but only run `MailerServices::TwoFactorAuthenticationRemovalRequestEmailer.new(publisher: publisher).perform` if the publisher exists

Comment (dlipeles) [reply app/controllers/publishers_controller.rb]: Thanks for pointing this out @yachtcaptain23 , I made the fix.

File: test/fixtures/two_factor_authentication_removals.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (yachtcaptain23) [inline test/fixtures/two_factor_authentication_removals.yml]: Unused fixture

File: test/models/two_factor_authentication_removal_test.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+require "test_helper"
+
+describe TwoFactorAuthenticationRemoval do
```
Comment (yachtcaptain23) [inline test/models/two_factor_authentication_removal_test.rb]: Would you add tests to make sure the before and after state are correct?

File: app/jobs/two_factor_authentication_removal_job.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      # First 2 weeks of the job, send reminder e-mails
+      if !two_factor_authentication_removal.two_factor_authentication_removal_time_completed?
+        MailerServices::TwoFactorAuthenticationRemovalReminderEmailer.new(publisher: publisher).perform
+
+      # 14th day of the job
+      elsif two_factor_authentication_removal.two_factor_authentication_removal_time_completed? && !two_factor_authentication_removal.removal_completed?
+        # Delete 2fa registrations, channels, and disconnect wallet
+        publisher.totp_registration.destroy if publisher.totp_registration.present?
+        publisher.u2f_registrations.destroy_all if !publisher.u2f_registrations.empty?
+        publisher.channels.destroy_all if !publisher.channels.empty?
```
Comment (yachtcaptain23) [inline app/jobs/two_factor_authentication_removal_job.rb]: Users need to still be paid. Please don't destroy, but you can remove the `publisher_id` on the `channels` table.

Comment (dlipeles) [reply app/jobs/two_factor_authentication_removal_job.rb]: Agreed, the users still need to, and will be paid; @evq's insight here was really helpful to me https://bravesoftware.slack.com/archives/C8GJLTQHW/p1552073068041900

The browser will hold funds until the publisher re-verifies their channels, for that reason we don't want channels to appear in the publishers list until they re-verify.

File: app/models/two_factor_authentication_removal.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    remainder = TOTAL_WAITING_PERIOD - (time - created_at)
+    if remainder > LOCKED_STATUS_WAITING_PERIOD
+      "Not started yet"
+    else
+      remainder = remainder / 24 / 3600
+      "#{remainder.round} Days"
+    end
+  end
+
+  def two_factor_authentication_removal_time_completed?
```
Comment (yachtcaptain23) [inline app/models/two_factor_authentication_removal.rb]: Prefer to bring this down to one line: 
`(Time.now - created_at) >= TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD`

File: app/models/two_factor_authentication_removal.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    time = Time.now
+    remainder = TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD - (time - created_at)
+    remainder <= 0
+  end
+
+  def locked_status_time_completed?
+    time = Time.now
```
Comment (yachtcaptain23) [inline app/models/two_factor_authentication_removal.rb]: Same here

File: app/models/two_factor_authentication_removal.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD = 1209600
+  # 4 Weeks represented in seconds
+  LOCKED_STATUS_WAITING_PERIOD = 2592000
+
+  # 6 Weeks represented in seconds
+  TOTAL_WAITING_PERIOD = TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD + LOCKED_STATUS_WAITING_PERIOD
+
+  # Returns time remaining in seconds
+  def total_time_remaining
+    time = Time.now
```
Comment (yachtcaptain23) [inline app/models/two_factor_authentication_removal.rb]: Prefer to be more ruby-like to simply be like `TOTAL_WAITING_PERIOD - (Time.now - created_at)`

File: app/services/mailer_services/two_factor_authentication_removal_request_emailer.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
+module MailerServices
+    class TwoFactorAuthenticationRemovalRequestEmailer < BaseService
```
Comment (yachtcaptain23) [inline app/services/mailer_services/two_factor_authentication_removal_request_emailer.rb]: Would you fix the indentation to be 2 spaces?

File: app/services/payout_report_publisher_includer.rb [hunk 9]
```diff
@@ -6,7 +6,8 @@ def initialize(payout_report:, publisher:, should_send_notifications:)
   end
 
   def perform
-    return if !@publisher.has_verified_channel? || @publisher.excluded_from_payout?
+    return if !@publisher.has_verified_channel? || @publisher.locked? || @publisher.excluded_from_payout?
+    publisher_has_unsettled_balance = false
```
Comment (yachtcaptain23) [inline app/services/payout_report_publisher_includer.rb]: Extra non-intended line?

File: app/jobs/two_factor_authentication_removal_job.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    two_factor_authentication_removals.each do |two_factor_authentication_removal|
+      publisher = Publisher.find(two_factor_authentication_removal.publisher_id)
+      # First 2 weeks of the job, send reminder e-mails
+      if !two_factor_authentication_removal.two_factor_authentication_removal_time_completed?
+        MailerServices::TwoFactorAuthenticationRemovalReminderEmailer.new(publisher: publisher).perform
+
+      # 14th day of the job
+      elsif two_factor_authentication_removal.two_factor_authentication_removal_time_completed? && !two_factor_authentication_removal.removal_completed?
+        # Delete 2fa registrations, channels, and disconnect wallet
+        publisher.totp_registration.destroy if publisher.totp_registration.present?
```
Comment (yachtcaptain23) [inline app/jobs/two_factor_authentication_removal_job.rb]: Would you lock this block into one transaction?

File: app/jobs/two_factor_authentication_removal_job.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        publisher.channels.destroy_all if !publisher.channels.empty?
+        publisher.disconnect_uphold
+        PublisherWalletDisconnector.new(publisher: publisher).perform
+        # Set status to Locked, mark 2fa removals as complete
+        publisher.status_updates.create(status: PublisherStatusUpdate::LOCKED)
+        two_factor_authentication_removal.update(removal_completed: true)
+
+      # After 6 weeks of the job. Restore publisher to Active state, delete the two_factor_authentication_removal object.
+      elsif two_factor_authentication_removal.locked_status_time_completed?
+        publisher.status_updates.create(status: PublisherStatusUpdate::ACTIVE)
```
Comment (yachtcaptain23) [inline app/jobs/two_factor_authentication_removal_job.rb]: One transaction here too

File: app/services/payout_report_publisher_includer.rb (hunk 9 above)
Comment (dlipeles) [reply app/services/payout_report_publisher_includer.rb]: Fixed, will keep some spacing for readability

**Comment by dlipeles** [comment]
Added the requested feedback, I also added a new `Security` tab on the admin dash for 2fa related information 

Screen Shot 2019-04-05 at 4 00 10 PM

File: app/jobs/two_factor_authentication_removal_job.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+      if !two_factor_authentication_removal.two_factor_authentication_removal_time_completed?
+        MailerServices::TwoFactorAuthenticationRemovalReminderEmailer.new(publisher: publisher).perform
+
+      # First waiting period completed - 14th Day
+      elsif two_factor_authentication_removal.two_factor_authentication_removal_time_completed? && !two_factor_authentication_removal.removal_completed?
+        # Remove 2fa registrations, remove channels, disconnect wallet, and put publisher into LOCKED state.
+        ActiveRecord::Base.transaction do
+          publisher.totp_registration.destroy if publisher.totp_registration.present?
+          publisher.u2f_registrations.destroy_all if !publisher.u2f_registrations.empty?
+          publisher.channels.destroy_all if !publisher.channels.empty?
```
Comment (yachtcaptain23) [inline app/jobs/two_factor_authentication_removal_job.rb]: Actually would you run `DeletePublisherChannelJob`? This makes a call to promo server to remove the unused channel.

File: app/controllers/publishers_controller.rb [hunk 13]
```diff
@@ -17,7 +17,12 @@ class PublishersController < ApplicationController
   ].freeze
 
   before_action :authenticate_via_token, only: %i(show)
-  before_action :authenticate_publisher!, except: %i(show)
+  before_action :authenticate_publisher!, except: %i(
+    show
+    two_factor_authentication_removal
```
Comment (corymcdonald) [inline app/controllers/publishers_controller.rb]: In my mind this is kind of clear indication that these methods likely belongs in another controller. This controller relates to a publisher and almost all of these methods require authentication via publisher. 

The reasoning I have for moving it into a new controller is
* they don't require any publisher to be logged in
* relates to a 2FA "model" as in everything about these can be mapped to create, and update

File: app/controllers/publishers_controller.rb [hunk 14]
```diff
@@ -109,6 +114,35 @@ def update
     end
   end
 
+  def request_two_factor_authentication_removal
+    publisher = Publisher.find_by_email(params[:email])
```
Comment (corymcdonald) [inline app/controllers/publishers_controller.rb]: Can you check to see if the case is sensitive on this?

File: app/controllers/publishers_controller.rb [hunk 15]
```diff
@@ -109,6 +114,35 @@ def update
...
+    if publisher
+      flash[:notice] = t("publishers.two_factor_authentication_removal.request_success")
+      MailerServices::TwoFactorAuthenticationRemovalRequestEmailer.new(publisher: publisher).perform
+    else
+      flash[:notice] = t("publishers.two_factor_authentication_removal.request_success")
```
Comment (corymcdonald) [inline app/controllers/publishers_controller.rb]: If the publisher doesn't exist we will say the same thing to the user saying that the request was success?

File: app/models/two_factor_authentication_removal.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
+class TwoFactorAuthenticationRemoval < ApplicationRecord
+  include ActionView::Helpers::DateHelper
+  belongs_to :publisher
+  # 2 Weeks represented in seconds
+  TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD = 1209600
```
Comment (corymcdonald) [inline app/models/two_factor_authentication_removal.rb]: Let's just do some ✨ Rails Magic 🌈 

`2.weeks.seconds`
`4.weeks.seconds`
`6.weeks.seconds`

File: app/models/two_factor_authentication_removal.rb [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    total_seconds = TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD - (Time.now - created_at)
+    distance_of_time_in_words(total_seconds)
+  end
+
+  def locked_status_days_remaining
+    total_seconds = TOTAL_WAITING_PERIOD - (Time.now - created_at)
+    if total_seconds <= LOCKED_STATUS_WAITING_PERIOD
+      distance_of_time_in_words(total_seconds)
+    else
+      "Not started yet"
```
Comment (corymcdonald) [inline app/models/two_factor_authentication_removal.rb]: Internationalization

File: app/services/payout_report_publisher_includer.rb [hunk 18]
```diff
@@ -6,10 +6,10 @@ def initialize(payout_report:, publisher:, should_send_notifications:)
   end
 
   def perform
-    return if !@publisher.has_verified_channel? || @publisher.excluded_from_payout?
+    return if !@publisher.has_verified_channel? || @publisher.locked? || @publisher.excluded_from_payout?
 
+    publisher_has_unsettled_balance = false
```
Comment (corymcdonald) [inline app/services/payout_report_publisher_includer.rb]: we assigning this twice?

File: config/routes.rb [hunk 19]
```diff
@@ -158,11 +161,13 @@
         patch :approve_channel
         get :statement
         post :create_note
+        get :cancel_two_factor_authentication_removal
```
Comment (corymcdonald) [inline config/routes.rb]: Really think these should just be routes on a `resources :two_factor_authentication_removal` controller

Comment (dlipeles) [reply config/routes.rb]: You are right, I will revisit this in a later PR and make this refactor

**Pull request merged**